File: static/typescript/src/seed/seed.ts
import pino from "pino";

import { DEFAULT_LANGUAGE, GREETINGS, MAX_NAME_LENGTH, MIN_NAME_LENGTH } from "./constants";
import { Language } from "./enums";
import { GreetingError } from "./errors";
import type { GreetingOptions, GreetingResult } from "./types";

const logger = pino({
  name: "seed",
  level: process.env["NODE_ENV"] === "test" ? "silent" : "info",
});
const supportedLanguages = new Set<string>(Object.values(Language));

export function isSupportedLanguage(value: string): value is Language {
  return value.length > 0 && supportedLanguages.has(value);
}

function validateName(name: string): void {
  if (name.length < MIN_NAME_LENGTH) {
    throw new GreetingError("Name is too short", "INVALID_NAME");
  }

  if (name.length > MAX_NAME_LENGTH) {
    throw new GreetingError("Name exceeds maximum length", "NAME_TOO_LONG");
  }

  if (name.trim() !== name) {
    throw new GreetingError("Name must not have leading or trailing whitespace", "INVALID_FORMAT");
  }
}

function resolveGreeting(language: Language): string {
  switch (language) {
    case Language.English:
    case Language.Spanish:
    case Language.French:
    case Language.German:
      return GREETINGS[language];
    default:
      throw new GreetingError(`Unsupported language: ${String(language)}`, "UNSUPPORTED_LANGUAGE");
  }
}

export function seed(name: string, options: GreetingOptions = {}): GreetingResult {
  const language = options.language ?? DEFAULT_LANGUAGE;
  validateName(name);

  const greeting = resolveGreeting(language);
  const result: GreetingResult = {
    greeting: `${greeting}, ${name}!`,
    language,
    name,
    timestamp: new Date(),
  };

  logger.info({ name, language }, "greeting generated");
  return result;
}

File: static/python/src/seed/seed.py
"""Greeting service with validation and structured logging."""

import logging

from .constants import DEFAULT_LANGUAGE, MAX_NAME_LENGTH
from .enums import Language
from .errors import SeedError
from .types import SeedResult

logger = logging.getLogger(__name__)


def greet(name: str, language: Language = DEFAULT_LANGUAGE) -> SeedResult:
    """Generate a greeting for a validated name."""
    validated_name = _validate_name(name)
    validated_language = _validate_language(language)
    message = _build_greeting(validated_name, validated_language)

    logger.info(
        "greeting generated",
        extra={
            "language": validated_language.value,
            "message_length": len(message),
            "name": validated_name,
        },
    )

    return {
        "message": message,
        "name": validated_name,
        "language": validated_language.value,
    }


def _validate_name(name: str) -> str:
    trimmed_name = name.strip()

    if not trimmed_name:
        raise SeedError("name must not be empty")

    if len(trimmed_name) > MAX_NAME_LENGTH:
        raise SeedError(f"name exceeds maximum length of {MAX_NAME_LENGTH}")

    return trimmed_name


def _validate_language(language: Language) -> Language:
    if not isinstance(language, Language):
        raise SeedError("unsupported language")

    return language


def _build_greeting(name: str, language: Language) -> str:
    greetings = {
        Language.ENGLISH: f"Hello, {name}!",
        Language.SPANISH: f"Hola, {name}!",
        Language.FRENCH: f"Bonjour, {name}!",
    }
    return greetings[language]
